src/synthetix/StakingRewards.t.sol

```diff
@@ -270,16 +285,115 @@ contract StakingRewardsTest is Test {
         assertEq(rewards.rewardsDuration(), 70 days);
     }
 
-    function testRevertSetRewardsDurationOnActiveDistribution() public {
-        setupStakingToken(100 * WAD);
-        rewards.stake(100 * WAD);
+    function testChangeRewardDurationOnActiveDistribution(uint256 newDuration) private {
+        uint256 totalStake = 100 * WAD;
+        uint256 totalReward = 123 * WAD;
+        init7DayFarm(totalStake, totalReward);
 
-        setupReward(100 * WAD);
+        skip(1 days);
+
+        rewards.setRewardsDuration(newDuration);
+
+        assertEq(rewards.rewardPerTokenStored(), 1 days * (totalReward / 7 days) * WAD / totalStake);
+        assertEq(rewards.lastUpdateTime(), block.timestamp);
+        assertEq(rewards.rewardRate(), 6 days * (totalReward / 7 days) / newDuration);
+        assertEq(rewards.periodFinish(), block.timestamp + newDuration);
+        assertEq(rewards.rewardsDuration(), newDuration);
 
+        skip(newDuration);
+
+        rewards.exit();
+
+        uint256 rewardLeft = totalReward % 7 days + 6 days * (totalReward / 7 days) % newDuration; // dust lost due to rewardRate being rounded down
+        uint256 rewardPaid = totalReward - rewardLeft;
+        assertEq(rewardGem.balanceOf(address(this)), rewardPaid);
+        assertEq(rewardGem.balanceOf(address(rewards)), rewardLeft);
+        assertEq(gem.balanceOf(address(this)), totalStake);
+        assertEq(gem.balanceOf(address(rewards)), 0);
+    }
+
+    function testIncreaseRewardsDurationOnActiveDistribution() public {
+        testChangeRewardDurationOnActiveDistribution(70 days);
+    }
+
+    function testDecreaseRewardsDurationOnActiveDistribution() public {
+        testChangeRewardDurationOnActiveDistribution(5 days);
+    }
+
+    function testSetSameRewardsDurationOnActiveDistribution() public {
+        uint256 totalStake = 100 * WAD;
+        uint256 totalReward = 123 * WAD;
+        init7DayFarm(totalStake, totalReward);
         skip(1 days);
 
-        vm.expectRevert("Previous rewards period must be complete before changing the duration for the new period");
-        rewards.setRewardsDuration(70 days);
+        uint256 prevRewardRate = rewards.rewardRate();
+        uint256 prevPeriodFinish = rewards.periodFinish();
+
+        rewards.setRewardsDuration(6 days); // same duration as time left
+
+        assertEq(rewards.rewardPerTokenStored(), 1 days * (totalReward / 7 days) * WAD / totalStake);
+        assertEq(rewards.lastUpdateTime(), block.timestamp);
+        assertEq(rewards.rewardRate(), prevRewardRate);
+        assertEq(rewards.periodFinish(), prevPeriodFinish);
+        assertEq(rewards.rewardsDuration(), 6 days);
+    }
+
+    function testSetRewardsDurationAndNotifyRewardAmountCallOrderOnActiveDistribution() public {
```

Comment:
There's a lot of repeated code in this test.
Couldn't we move all the assertions, `skip` + `rewards.exit()` + `vm.revertTo` to a helper function?

Reply:
Prefer to leave as-is to keep the test self-contained and avoid using too many external helpers.